### ISSUE TITLE
Support for async rendering

### ISSUE THREAD

**Issue by SebastianSchenk** [angular/angular#19030]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

The new version of React (16.x) will introduce a core feature called async rendering.
(https://edgecoders.com/react-16-features-and-fiber-explanation-e779544bb1b7)

This allows the framework to render a lot of things at the same time without blocking the main thread.

I was wondering, is this something that can also be implemented in angular? If so, are there any plans for a similar feature?

## Current behavior

Component rendering is happening synchronous --> blocks the main thread.

## Expected behavior

Component rendering is happening asynchronous --> doesn't block the main thread.

**Comment by mlc-mlapis** [comment]
> **React Async Rendering**
> The focus of the initial 16.0 release is on compatibility for existing applications. Async rendering will not be an option initially, but in later 16.x releases, it will be included as an opt-in feature.

Web Worker solution for async rendering is still a big question. There are no serious data to compare the final effects. It'll be probably highly dependent on exact cases.

**Comment by SebastianSchenk** [comment]
@mlc-mlapis yeah I know it is not included yet, but it will be.

> You can run whatever code you like inside the worker thread, with some exceptions. For example, you can't directly manipulate the DOM from inside a worker, or use some default methods and properties of the window object. But you can use a large number of items available under window, including WebSockets, and data storage mechanisms like IndexedDB and the Firefox OS-only Data Store API. See Functions and classes available to workers for more details.
https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers

I don't think web workers will effect render performance in the same way, since you cannot access the DOM there. I think the idea behind web workers is to perform heavy computation on a seperat thread.

**Comment by mlc-mlapis** [comment]
@SebastianSchenk ... yes, it's true ... `to perform heavy computation on a seperat thread`. These type of tasks are available even today for any app ... including Angular.

There are Angular modules:
```
"@angular/platform-webworker": "4.3.6",
"@angular/platform-webworker-dynamic": "4.3.6"
```

You can also look at:
https://blog.angularindepth.com/angular-with-web-workers-step-by-step-dc11d5872135
https://github.com/Owain94/angular-webworker

**Comment by gund** [comment]
Out of curiosity I've ported Fiber vs Stack react demo to Angular [here](https://gund.github.io/ng-s-triangle-demo/).
I thought that app in webworker will lead to significant performance difference but actually animations still was junky because worker thread was not able to keep up with 60 fps.
However main thread clearly was not blocked.

**Comment by mlc-mlapis** [comment]
@gund ... both links are not working ... bundles missing ... 404.
https://gund.github.io/ng-s-triangle-demo/single
https://gund.github.io/ng-s-triangle-demo/worker

**Comment by gund** [comment]
@mlc-mlapis thanks for pointing out, just base tag problem.
Fixed now =)

**Comment by Toxicable** [comment]
@gund interesting, thanks for the demos.
just to add to it, if you move your viewport to the size where you demo makes a scroll bar you can tell that the work is being done on a background thread by being able to scroll while it's moving still, unlike the sync demo.

**Comment by gund** [comment]
@Toxicable yeah the fact that main thread was not blocked is really good, but if you were doing animations from the worker thread and it was blocked then still visually user might have a perception of a bad performance.
Clearly async rendering is about smart scheduling and prioritizing operations to DOM...

**Comment by Toxicable** [comment]
As far as I understand it, while running your app inside a Web Worker would be the best performance possible the bus between Worker and Main thread is too slow to get reasonable performance boosts from using it

**Comment by mlc-mlapis** [comment]
@Toxicable ... yeap, the same observation as @ghetolay observed some time ago ... the bus is the limitation. So cases where Web Worker threads are optimal to use are those ones where a lot of processing is required inside Web Worker thread with less frequencies of communication back to the main thread.

**Comment by gund** [comment]
I really do not think it's about bus throughput.
It looks like the worker thread simply unable to compute fast enough to push animation updates to UI thread.
If you will remove slow code it probably will result in 60 fps animation in worker mode same as in single thread mode.

**Comment by mlc-mlapis** [comment]
@gund ... hmm, as I remember the best performance is possible to get with `stringifying` of everything because the transfer of strings is 2-3x faster then objects.

Also for 60  frames per second it is necessary that no JS runs longer then 16 ms ... as https://hacks.mozilla.org/2015/07/how-fast-are-web-workers

**Comment by gund** [comment]
I just tried locally webworker version but without heavy load on worker thread and as expected I got 60 fps silky-smooth animations on UI.
So there is really nothing to do with webworkers - it's purely scheduling and prioritization problem...

**Comment by aretheregods** [comment]
@SebastianSchenk I just wanted to correct the statement that 

> The new version of React (16.x) will introduce a core feature called async rendering.

It doesn't include async rendering at the moment, and it's unclear _when it will_. Right now, the React team says [later this year.](https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html) But, as it stands React and Angular are about equally fast (with Angular being a little faster) according to [this benchmark](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts-results/table.html)

**Comment by shyamal890** [comment]
It seems reconciliation should already be available in Angular not sure if any thing changed in last 8 months since the issue was created.

**Comment by hiepxanh** [comment]
I see that stencil with demo is can do that [stencil by ionic demo website: smooth ](https://stencil-fiber-demo.firebaseapp.com/). I tried with service-worker but no luck. 

[github repository](https://github.com/hiepxanh/Ng6-Sierpinski-triangle/tree/gh-pages)
[ng6 demo website: nope](https://hiepxanh.github.io/Ng6-Sierpinski-triangle/)
[ng4 demo website: nope!](https://gund.github.io/ng-s-triangle-demo/)
[react demo website: smooth](https://claudiopro.github.io/react-fiber-vs-stack-demo/)

**Comment by splincode** [comment]
@hiepxanh You realized a bad implementation

@trotyl @Toxicable @IgorMinar @mhevery I just added requestAnimationFrame + requestIdleCallback and received 60fps

![image](https://user-images.githubusercontent.com/12021443/44015370-4eb4b2ce-9ed9-11e8-8ef4-7b7052588942.png)

Angular 6:
https://angular6-store.stackblitz.io/

React Fiber:
https://claudiopro.github.io/react-fiber-vs-stack-demo/fiber.html

**Comment by hiepxanh** [comment]
wow, that amazing, thank you. how can you do that?

**Comment by splincode** [comment]
@hiepxanh You can see
https://stackblitz.com/edit/angular6-store

**Comment by hiepxanh** [comment]
thank you sir. that good

**Comment by Ploppy3** [comment]
@splincode Even though the perf improvement is quite noticeable, the Angular app still stutters on every change detection cycle, could this be improved or is it just that Angular is slower? I'm even more concerned since I noticed that [Angular Ivy is slower](https://github.com/angular/angular/issues/21706#issuecomment-438444323) in its current state than the latest stable version of Angular.

**Comment by stupidawesome** [comment]
[Challenge accepted](https://stackblitz.com/edit/angular-25scyj). Here are the unscientific results.

![image](https://user-images.githubusercontent.com/31086926/58410665-92acb380-80ad-11e9-8790-a1149da4f90f.png)

Included in the screen are the previous two examples, plus [another example](https://radi.js.org/perf-test.html) I came across that's even faster.

The key to 60fps is to not run change detection more than you need to. I wrote a library called [NgObservable](https://github.com/zodiac-team/zodiac-ui/tree/master/libs/ng-observable) that brings a different change detection mode to the table called `DETACH`. In this mode views are detached from their parents and change detection only runs when you tell it to.

**Comment by helmbold** [comment]
Since React is mentioned here, React [Concurrent Mode](https://reactjs.org/docs/concurrent-mode-intro.html) is probably of interest in the context of this issue.

**Comment by Ploppy3** [comment]
Bumping so that this issue is not aggressively closed.

**Comment by DmitryEfimenko** [comment]
2021 bump so that it does not get closed

**Comment by hiepxanh** [comment]
@IgorMinar  sir, can you give a few explain 👍

**Comment by mmuscat** [comment]
Detached change detection gets you close to 60fps without async rendering, which is probably enough for real world applications that aren't updating 1500+ nodes in an animation frame. More importantly, the UI is still responsive because Angular can check a single view without traversing and re-rendering the entire node tree on every change.

https://mmuscat.github.io/striangle-demo/

Other demos in this thread seem to have problems on high refresh rate screens or when dev tools are open. The timer drifts, skips or stops entirely. The stencil demo seems to work best by splitting node updates into chunks across several frames, which you'll notice as the page first renders. I'd prefer having UI consistency and looking at more granular change detection to remove the cost of redundant checks.
